src/certgen_certificate_generator_server.cc

```diff
@@ -27,11 +27,285 @@
+            if (rv != 0)
+            {
+                log_error ("Error while getting hostname for IP address %s %s", ipv6Address.c_str(), gai_strerror (rv));
+                continue;
+            }
+
+            auto it = std::find (result.begin(), result.end(), name);
+            if (it == result.end())
+            {
+                result.push_back (name);
+            }
+        }
+
+        return result;
+    }
+
+    std::list<std::string> CertificateGeneratorServer::getSystemIPs(int paramFamily)
+    {
+        struct ifaddrs *interfaces, *iface;
+        char host[NI_MAXHOST];
+        std::list<std::string> result;
+
+        if (getifaddrs (&interfaces) == -1)
+        {
+            throw std::runtime_error ("Unable to get IP adresses");
+        }
+        iface = interfaces;
+        for (iface = interfaces; iface != NULL; iface = iface->ifa_next)
+        {
+            if (iface->ifa_addr == NULL) continue;
+            int family = iface->ifa_addr->sa_family;
+            if (family == paramFamily)
+            {
+                if ( getnameinfo(iface->ifa_addr,
+                        (family == AF_INET) ? sizeof(struct sockaddr_in) :
+                        sizeof(struct sockaddr_in6),
+                        host, NI_MAXHOST,
+                        NULL, 0, NI_NUMERICHOST) == 0
+                   )
+                {
+                    // sometimes IPv6 addres looks like ::2342%IfaceName
+                    char *p = strchr (host, '%');
+                    if (p) *p = 0;
+
+                    auto it = std::find (result.begin(), result.end(), host);
+                    if (it == result.end())
+                    {
+                        result.push_back (host);
+                    }
+                }
+            }
+        }
+        freeifaddrs (interfaces);
+        return result;
+    }
+
+
+    std::string CertificateGeneratorServer::handleGenerateSelfsignedCertificate(const fty::Payload & params)
+    {
+        // load config file for the service
+        CertificateGeneratorConfig certgenConfig;
+
+        if (params.empty() || params[0].empty ())
+        {
+            throw std::runtime_error ("Missing service name");
+        }
+
+        std::string serviceName (params[0]);
+        std::string configFilePath(CONFIG_PATH + serviceName + ".cfg");
+        std::ifstream configFile (configFilePath);
+        if (!configFile)
+        {
+            throw std::runtime_error ("Could not open file " + configFilePath);
+        }
+
+        configFile.exceptions ( std::ifstream::failbit | std::ifstream::badbit );
+        std::stringstream configJson;
+        configJson << configFile.rdbuf();
+        configFile.close ();
+
+        cxxtools::SerializationInfo certgenSi;
+        cxxtools::JsonDeserializer deserializer (configJson);
+        deserializer.deserialize (certgenSi);
+
+        certgenSi >>= certgenConfig;
+
+        fty::CertificateConfig config;
```

Comment:
create static function for filling out the CertificateConfig

Reply:
Addressed.